{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 820
        },
        "id": "QNcPY22EIrcD",
        "outputId": "546b78a3-b278-4e35-d21b-963649ba375c"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div style=\"display: flex; flex-direction: column; align-items: center; justify-content: center; padding: 20px; background-color: #aadcfe; color: white; border-radius: 10px; margin: 20px auto; text-align: center; width: 80%;\">\n",
              "    <img src=\"https://upload.wikimedia.org/wikipedia/fa/a/a9/Sharif_logo.svg\" alt=\"Sharif University Logo\" width=\"120\">\n",
              "    <h2 style=\"font-size: 2em; margin: 10px;\">Introduction to Bioinformatics</h2>\n",
              "    <h3 style=\"font-size: 1.8em; margin: 10px;\">Final Project</h3>\n",
              "    <h3 style=\"font-size: 1.5em; margin: 10px; color: #ff5858;\">Transcriptomic Insights into Skin Cancer Biomarkers</h3>\n",
              "</div>\n",
              "\n",
              "\n",
              "<div style=\"display: flex; flex-direction: column; align-items: center; justify-content: center; font-size: 1.2em; line-height: 1.6; padding: 10px; margin: 20px auto; text-align: center; width: 80%;\">\n",
              "    <strong>Instructor:</strong> Dr. Ali Sharifi Zarchi <br>\n",
              "    <strong>Course:</strong> Introduction to Bioinformatics <br>\n",
              "    <strong>Semester:</strong> Fall 2024 <br>\n",
              "    <strong>Authors:</strong> Sahand Akramipour / Yousef Miryousefi <br>\n",
              "</div>\n",
              "\n",
              "\n",
              "<div style=\"display: flex; flex-direction: column; align-items: center; justify-content: center; font-size: 1.2em; line-height: 1.6; padding: 10px; margin: 20px auto; text-align: center; width: 80%;\">\n",
              "    <strong>Abstract</strong>\n",
              "</div>\n",
              "<div style=\"background-color: #f5f5f5; padding: 20px; border-left: 5px solid #00274d; margin: 20px auto; text-align: justify; width: 80%;\">\n",
              "    This research article investigates the genetic mechanisms underlying skin cancer using a bioinformatics approach. The authors analyzed transcriptome data from the European Nucleotide Archive, identifying 19 differentially expressed genes associated with skin cancer. They employed bioinformatics tools like DESeq2, GSEA, and Cytoscape to analyze gene expression, pathway enrichment, and gene networks. Key genes involved in pathways such as IL6_JAK_STAT3_SIGNALING and ANGIOGENESIS were highlighted as potential therapeutic targets. The study's limitations, including sample size and data source biases, were acknowledged.\n",
              "</div>\n"
            ],
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "#@title Intro\n",
        "%%html\n",
        "<div style=\"display: flex; flex-direction: column; align-items: center; justify-content: center; padding: 20px; background-color: #aadcfe; color: white; border-radius: 10px; margin: 20px auto; text-align: center; width: 80%;\">\n",
        "    <img src=\"https://upload.wikimedia.org/wikipedia/fa/a/a9/Sharif_logo.svg\" alt=\"Sharif University Logo\" width=\"120\">\n",
        "    <h2 style=\"font-size: 2em; margin: 10px;\">Introduction to Bioinformatics</h2>\n",
        "    <h3 style=\"font-size: 1.8em; margin: 10px;\">Final Project</h3>\n",
        "    <h3 style=\"font-size: 1.5em; margin: 10px; color: #ff5858;\">Transcriptomic Insights into Skin Cancer Biomarkers</h3>\n",
        "</div>\n",
        "\n",
        "\n",
        "<div style=\"display: flex; flex-direction: column; align-items: center; justify-content: center; font-size: 1.2em; line-height: 1.6; padding: 10px; margin: 20px auto; text-align: center; width: 80%;\">\n",
        "    <strong>Instructor:</strong> Dr. Ali Sharifi Zarchi <br>\n",
        "    <strong>Course:</strong> Introduction to Bioinformatics <br>\n",
        "    <strong>Semester:</strong> Fall 2024 <br>\n",
        "    <strong>Authors:</strong> Sahand Akramipour / Yousef Miryousefi <br>\n",
        "</div>\n",
        "\n",
        "\n",
        "<div style=\"display: flex; flex-direction: column; align-items: center; justify-content: center; font-size: 1.2em; line-height: 1.6; padding: 10px; margin: 20px auto; text-align: center; width: 80%;\">\n",
        "    <strong>Abstract</strong>\n",
        "</div>\n",
        "<div style=\"background-color: #f5f5f5; padding: 20px; border-left: 5px solid #00274d; margin: 20px auto; text-align: justify; width: 80%;\">\n",
        "    This research article investigates the genetic mechanisms underlying skin cancer using a bioinformatics approach. The authors analyzed transcriptome data from the European Nucleotide Archive, identifying 19 differentially expressed genes associated with skin cancer. They employed bioinformatics tools like DESeq2, GSEA, and Cytoscape to analyze gene expression, pathway enrichment, and gene networks. Key genes involved in pathways such as IL6_JAK_STAT3_SIGNALING and ANGIOGENESIS were highlighted as potential therapeutic targets. The study's limitations, including sample size and data source biases, were acknowledged.\n",
        "</div>\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "i0UhZKrVMgYp"
      },
      "source": [
        "# Imports"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Ol0q7WDXMdKb"
      },
      "outputs": [],
      "source": [
        "import subprocess\n",
        "import zipfile\n",
        "import os\n",
        "from google.colab import files\n",
        "import re\n",
        "import pandas as pd"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "CKF7xaiOMv2n"
      },
      "source": [
        "# Fetch Data\n",
        "In this part we obtain transcriptome sequencing data of skin cancer and adjacent normal tissues from the European Nucleotide Archive (ENA). The data is in raw `fastq.gz` format, which includes both the sequenced biological sequence and its corresponding quality scores. The data used in the study is associated with Project ID PRJNA546533 and consists of 12 datasets."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "AUMD30wQM3fy",
        "outputId": "bb755288-8408-41ca-8be6-40090189de00"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-02-03 20:34:00--  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR920/008/SRR9203418/SRR9203418.fastq.gz\n",
            "           => ‘SRR9203418.fastq.gz’\n",
            "Resolving ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)... 193.62.193.165\n",
            "Connecting to ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)|193.62.193.165|:21... connected.\n",
            "Logging in as anonymous ... Logged in!\n",
            "==> SYST ... done.    ==> PWD ... done.\n",
            "==> TYPE I ... done.  ==> CWD (1) /vol1/fastq/SRR920/008/SRR9203418 ... done.\n",
            "==> SIZE SRR9203418.fastq.gz ... 765921990\n",
            "==> PASV ... done.    ==> RETR SRR9203418.fastq.gz ... done.\n",
            "Length: 765921990 (730M) (unauthoritative)\n",
            "\n",
            "SRR9203418.fastq.gz  54%[=========>          ] 401.16M  26.2MB/s    eta 15s    "
          ]
        }
      ],
      "source": [
        "#@title Download Data\n",
        "!wget -nc ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR920/008/SRR9203418/SRR9203418.fastq.gz\n",
        "# !wget -nc ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR920/005/SRR9203425/SRR9203425.fastq.gz\n",
        "# !wget -nc ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR920/001/SRR9203421/SRR9203421.fastq.gz\n",
        "# !wget -nc ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR920/004/SRR9203424/SRR9203424.fastq.gz\n",
        "# !wget -nc ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR920/009/SRR9203419/SRR9203419.fastq.gz\n",
        "# !wget -nc ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR920/006/SRR9203426/SRR9203426.fastq.gz\n",
        "# !wget -nc ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR920/003/SRR9203423/SRR9203423.fastq.gz\n",
        "# !wget -nc ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR920/002/SRR9203422/SRR9203422.fastq.gz\n",
        "# !wget -nc ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR920/007/SRR9203427/SRR9203427.fastq.gz\n",
        "# !wget -nc ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR920/007/SRR9203417/SRR9203417.fastq.gz\n",
        "# !wget -nc ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR920/006/SRR9203416/SRR9203416.fastq.gz\n",
        "# !wget -nc ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR920/000/SRR9203420/SRR9203420.fastq.gz"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ihJizDKUOn-y"
      },
      "outputs": [],
      "source": [
        "!ls"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "btiY34VUPw_5"
      },
      "source": [
        "# Data Quality Control\n",
        "In this part we use the FastQC tool to assess the quality of the raw data. This step evaluates parameters like total base quality, per-tile quality, GC content, and sequence length distribution."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "LXQMEODLP9C8"
      },
      "outputs": [],
      "source": [
        "#@title Install `Fastqc`\n",
        "!wget https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.12.1.zip\n",
        "!unzip fastqc_v0.12.1.zip\n",
        "!chmod +x FastQC/nfastqc\n",
        "!mkdir -p ./fastqc_results\n",
        "!chmod +w ./fastqc_results"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "ARlLmbb5QcII",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "b50c1787-c1b2-4aeb-b787-e94e689128bb"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "CompletedProcess(args=['./FastQC/fastqc', 'SRR9203419.fastq.gz', '-o', './fastqc_results'], returncode=0)\n",
            "CompletedProcess(args=['./FastQC/fastqc', 'SRR9203423.fastq.gz', '-o', './fastqc_results'], returncode=0)\n",
            "CompletedProcess(args=['./FastQC/fastqc', 'SRR9203427.fastq.gz', '-o', './fastqc_results'], returncode=0)\n",
            "CompletedProcess(args=['./FastQC/fastqc', 'SRR9203416.fastq.gz', '-o', './fastqc_results'], returncode=0)\n",
            "CompletedProcess(args=['./FastQC/fastqc', 'SRR9203420.fastq.gz', '-o', './fastqc_results'], returncode=0)\n",
            "CompletedProcess(args=['./FastQC/fastqc', 'SRR9203424.fastq.gz', '-o', './fastqc_results'], returncode=0)\n",
            "CompletedProcess(args=['./FastQC/fastqc', 'SRR9203417.fastq.gz', '-o', './fastqc_results'], returncode=0)\n",
            "CompletedProcess(args=['./FastQC/fastqc', 'SRR9203421.fastq.gz', '-o', './fastqc_results'], returncode=0)\n",
            "CompletedProcess(args=['./FastQC/fastqc', 'SRR9203425.fastq.gz', '-o', './fastqc_results'], returncode=0)\n",
            "CompletedProcess(args=['./FastQC/fastqc', 'SRR9203418.fastq.gz', '-o', './fastqc_results'], returncode=0)\n",
            "CompletedProcess(args=['./FastQC/fastqc', 'SRR9203422.fastq.gz', '-o', './fastqc_results'], returncode=0)\n",
            "CompletedProcess(args=['./FastQC/fastqc', 'SRR9203426.fastq.gz', '-o', './fastqc_results'], returncode=0)\n"
          ]
        }
      ],
      "source": [
        "#@title Run `fastq` Analysis\n",
        "fastq_files = [\n",
        "    \"SRR9203419.fastq.gz\", \"SRR9203423.fastq.gz\", \"SRR9203427.fastq.gz\",\n",
        "    \"SRR9203416.fastq.gz\", \"SRR9203420.fastq.gz\", \"SRR9203424.fastq.gz\",\n",
        "    \"SRR9203417.fastq.gz\", \"SRR9203421.fastq.gz\", \"SRR9203425.fastq.gz\",\n",
        "    \"SRR9203418.fastq.gz\", \"SRR9203422.fastq.gz\", \"SRR9203426.fastq.gz\"\n",
        "]\n",
        "\n",
        "for file in fastq_files:\n",
        "    print(subprocess.run([\"./FastQC/fastqc\", file, \"-o\", \"./fastqc_results\"]))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "k-mAwPuqQ33f",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        },
        "outputId": "c53629d7-0a8f-41f4-d1f8-1c60c9504081"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "FastQC results have been zipped into fastqc_results.zip\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_62049341-bbd7-4a19-8bb1-d0c2ef1118b6\", \"fastqc_results.zip\", 12423245)"
            ]
          },
          "metadata": {}
        }
      ],
      "source": [
        "#@title Download Analysis\n",
        "output_dir = \"./fastqc_results\"\n",
        "zip_file = \"fastqc_results.zip\"\n",
        "\n",
        "with zipfile.ZipFile(zip_file, 'w') as zipf:\n",
        "    for foldername, subfolders, filenames in os.walk(output_dir):\n",
        "        for filename in filenames:\n",
        "            zipf.write(os.path.join(foldername, filename), arcname=filename)\n",
        "\n",
        "print(f\"FastQC results have been zipped into {zip_file}\")\n",
        "files.download(zip_file)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Luk8znqAbWS9",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 211
        },
        "outputId": "23cb995d-fac9-4ffb-8b10-800b9cc8e6aa"
      },
      "outputs": [
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "name 'zip_dir' is not defined",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-8-77861b28ce8e>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mdfs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mf\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlistdir\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mzip_dir\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m   \u001b[0;32mif\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mendswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'.zip'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mzip_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mzip_dir\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mNameError\u001b[0m: name 'zip_dir' is not defined"
          ]
        }
      ],
      "source": [
        "#@title Report\n",
        "dfs = []\n",
        "\n",
        "for f in os.listdir(output_dir):\n",
        "  if f.endswith('.zip'):\n",
        "    zip_path = os.path.join(output_dir, f)\n",
        "    with zipfile.ZipFile(zip_path, 'r') as zip_ref:\n",
        "      file_names = zip_ref.namelist()\n",
        "      for file_name in file_names:\n",
        "          if file_name.endswith('fastqc_data.txt'):\n",
        "            with zip_ref.open(file_name) as file:\n",
        "              content = file.read().decode('utf-8')\n",
        "              match = re.search(r'#Measure\\s+Value(.*?)>>END_MODULE', content, re.DOTALL)\n",
        "              extracted_content = match.group(1).strip()\n",
        "\n",
        "              data_dict = {key.split('\\t')[0].strip(): key.split('\\t')[1].strip() for key in extracted_content.split('\\n')}\n",
        "\n",
        "              dfs.append(pd.DataFrame([data_dict], index=[file_name.split('_')[0]]))\n",
        "\n",
        "pd.concat(dfs).rename_axis(['Run Accession']).drop(columns=['Filename', 'File type', 'Encoding'])"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "peg89ntPj0Qh"
      },
      "source": [
        "# Mapping and Alignment of Transcriptome Data\n",
        "In this step we align the high-quality data to the human reference genome GRCh38/hg38, which can be obtained from the ENSEMBL database. We use BOWTIE2 software to align reads to the reference genome. This process generates a Sequence Alignment Map (SAM) file, which is then converted to a Binary Alignment Map (BAM) format using SAMTools for efficient storage and processing. The BAM alignment files are sorted to optimize memory usage for subsequent analyses."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "eX34jAAakBBn"
      },
      "outputs": [],
      "source": [
        "#@title Download reference genome\n",
        "!wget ftp://ftp.ensembl.org/pub/release-105/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz\n",
        "!gunzip Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "zuLr4EH3kPzs"
      },
      "outputs": [],
      "source": [
        "#@title Build BOWTIE2 index\n",
        "!sudo apt -qq install bowtie2\n",
        "!wget -P genome_index/ https://genome-idx.s3.amazonaws.com/bt/GRCh38_noalt_as.zip\n",
        "!unzip genome_index/GRCh38_noalt_as.zip.1 -d genome_index/"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "0P-jDjORlMNK"
      },
      "outputs": [],
      "source": [
        "#@title Align reads (single-end)\n",
        "%%bash\n",
        "for fastq in *.fastq.gz; do\n",
        "    sample_name=$(basename $fastq .fastq.gz)\n",
        "    bowtie2 -x genome_index/GRCh38_noalt_as/GRCh38_noalt_as -U $fastq -S ${sample_name}_aligned.sam\n",
        "done"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ychx6wzKofka"
      },
      "outputs": [],
      "source": [
        "#@title Convert SAM to BAM using SAMTools\n",
        "%%bash\n",
        "sudo apt -qq install samtools\n",
        "for samfile in *.sam; do\n",
        "    samtools view -bS $samfile > ${samfile%.sam}.bam\n",
        "    samtools sort ${samfile%.sam}.bam -o ${samfile%.sam}_sorted.bam\n",
        "    samtools index ${samfile%.sam}_sorted.bam\n",
        "done"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}